overlord/hookstate/ctlcmd/refresh.go

```diff
@@ -224,9 +224,18 @@ func (c *refreshCommand) hold() error {
 	// cache the action so that hook handler can implement default behavior
 	ctx.Cache("action", snapstate.GateAutoRefreshHold)
 
-	var affecting []string
-	if err := ctx.Get("affecting-snaps", &affecting); err != nil {
-		return fmt.Errorf("internal error: cannot get affecting-snaps")
+	affected, err := snapstate.AffectedByRefreshCandidates(st)
+	if err != nil {
+		return err
+	}
+
+	affectedInfo := affected[ctx.InstanceName()]
+	if affectedInfo == nil {
+		return fmt.Errorf("no snaps are held by %q", ctx.InstanceName())
```

Comment:
what happened about this before the change? is this a new error? would this just be ignored? it merits a dedicated test I suppose

Reply:
Previously (with "affecting-snaps" stored in the hook context) we would not have this scenario here because the list was pre-computed early and wouldn't be empty. But now since AffectedByRefreshCandidates() re-computes this, in theory this might be the case and would signify an internal error (meaning we handled conflicts wrong and something changed while we are running this change and the hook). I'll change this to internal error and add a test.

Reply:
Done.

Reply:
this is also mostly impossible because an updating snap affects itself ?

Reply:
I think it is still theoretically possible in case of a bug I explained above; that snap itself may not be updated but it's affected by another snap and have its hook executed; however I've reworded it to ""internal error: snap %q is not affected by any snaps".